### PR TITLE
Rulefit: do not remove bad columns

### PR DESCRIPTION
**Pull request by koniecsveta**
This fixes the bug on rulefit during benchmark test case with springleaf dataset.

Each tree node remembers column id of the feature it is splitting on.
When constructing condition which is representing a tree node, this column id is used as condition's feature id.
When creating GLM's training frame this feature id is used during condition evaluation on the training set rows to map the result to correct place in chunk.
When in between of this we drop some bad columns, mapping between training frame and Condition#featureIndex/Node#colId gets broken and I have IndexOutOfBoundsException.

### CODE REVIEW

File: h2o-algos/src/main/java/hex/rulefit/RuleFit.java [hunk 1]
```diff
@@ -90,6 +90,11 @@ public void init(boolean expensive) {
         // if (expensive && error_count() == 0) checkMemoryFootPrint();
     }
 
+    @Override
+    protected void ignoreBadColumns(int npredictors, boolean expensive) {
+        // Do not remove bad columns - we need them for correct GLM train frame creation
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/RuleFit.java:95]: this does not seem right, I think you might want to remove bad/constant cols at the beginning of RuleFit

**Review by honzasterba — COMMENTED** [review]
also adding a test case for this edge-case would be a good idea

File: h2o-algos/src/main/java/hex/rulefit/RuleFit.java (hunk 1 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/rulefit/RuleFit.java:95]: I think the fact that this fixes an issue might be a sign of something bad happening

@koniecsveta how are frames passed to glm and drf/gbm? do you use trainModelNested?

Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/RuleFit.java:95]: no no I don't use trainModelNested and removing bad cols at the beginning didn't help

Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/RuleFit.java:95]: like this, it still goes into ModelBuilder#ignoreBadCols() on glm's and drf/gbm's model builder level, by this I just avoid going there on rulefit's model builder level

File: h2o-algos/src/main/java/hex/rulefit/RuleFit.java [hunk 2]
```diff
@@ -181,14 +183,17 @@ public void computeImpl() {
             try {
                 // linearTrain = frame to be used as _train for GLM in 2., will be filled in 1.
                 Frame linearTrain = new Frame(Key.make("paths_frame" + _result));
-                
+                // store train frame without bad columns to pass it to tree model builders
+                Frame trainRebalanced = new Frame(_train);
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/rulefit/RuleFit.java:187]: I don't think "rebalanced" is a good name - it implies the dataset was rebalanced which is not always the case (rebalancing means redistributing the chunks)

it is (from RuleFit's point of view) - "adapted frame"

Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/RuleFit.java]: 5bd6d19

**Review by honzasterba — APPROVED** [review]

**Review by michalkurka — APPROVED** [review]
